src/coreclr/jit/gentree.cpp

```diff
@@ -24133,6 +24169,559 @@ GenTree* Compiler::gtNewSimdSumNode(var_types type, GenTree* op1, CorInfoType si
 #endif // !TARGET_XARCH && !TARGET_ARM64
 }
 
+#if defined(TARGET_XARCH)
+GenTree* Compiler::gtNewSimdTernaryLogicNode(var_types   type,
+                                             GenTree*    op1,
+                                             GenTree*    op2,
+                                             GenTree*    op3,
+                                             GenTree*    op4,
+                                             CorInfoType simdBaseJitType,
+                                             unsigned    simdSize)
+{
+    assert(IsBaselineVector512IsaSupportedDebugOnly());
+
+    assert(varTypeIsSIMD(type));
+    assert(getSIMDTypeForSize(simdSize) == type);
+
+    assert(op1 != nullptr);
+    assert(op1->TypeIs(type));
+
+    assert(op2 != nullptr);
+    assert(op2->TypeIs(type));
+
+    assert(op3 != nullptr);
+    assert(op3->TypeIs(type));
+
+    assert(op4 != nullptr);
+    assert(genActualType(op4) == TYP_INT);
+
+    var_types simdBaseType = JitType2PreciseVarType(simdBaseJitType);
+    assert(varTypeIsArithmetic(simdBaseType));
+
+    NamedIntrinsic intrinsic = NI_Illegal;
+
+    if (simdSize == 64)
+    {
+        intrinsic = NI_AVX512F_TernaryLogic;
+    }
+    else
+    {
+        assert((simdSize == 16) || (simdSize == 32));
+        intrinsic = NI_AVX512F_VL_TernaryLogic;
+    }
+
+    if (op4->IsIntegralConst())
+    {
+        // We have a constant control byte, so we can potentially optimize
+
+        uint8_t                 control  = static_cast<uint8_t>(op4->AsIntCon()->gtIconVal);
+        const TernaryLogicInfo& info     = TernaryLogicInfo::lookup(control);
+        TernaryLogicUseFlags    useFlags = info.GetAllUseFlags();
+
+        if (useFlags != TernaryLogicUseFlags::ABC)
+        {
+            // We are not using all 3 inputs, so we can potentially optimize
+
+            assert(info.oper2 != TernaryLogicOperKind::Select);
+            assert(info.oper2 != TernaryLogicOperKind::True);
+            assert(info.oper2 != TernaryLogicOperKind::False);
+            assert(info.oper2 != TernaryLogicOperKind::Cond);
+            assert(info.oper2 != TernaryLogicOperKind::Major);
+            assert(info.oper2 != TernaryLogicOperKind::Minor);
+            assert(info.oper3 == TernaryLogicOperKind::None);
+            assert(info.oper3Use == TernaryLogicUseFlags::None);
+
+            GenTree* value1;
+            GenTree* value2;
+
+            switch (useFlags)
+            {
+                case TernaryLogicUseFlags::A:
+                {
+                    // We'll swap from 'A, B, C' to 'B, C, A'
+                    // We expect A to have been spilled
+
+                    assert((op1->gtFlags & GTF_SIDE_EFFECT) == 0);
```

Comment:
This is not a sufficient assert to make reordering ok, `A` can also have global references or ordering side effects that disallows it from being reordered with B and C. There's no precise way to check for this during import, only super conservative `gtHasLocalsWithAddrOp`.

Reply:
I have import spilling these up front. Is there a better check that can be done for post-import?